actors/builtin/miner/miner_actor.go

```diff
@@ -2067,6 +2068,253 @@ func (a Actor) RepayDebt(rt Runtime, _ *abi.EmptyValue) *abi.EmptyValue {
 }
 
+type ReplicaUpdate struct {
+	SectorID           abi.SectorNumber
+	Deadline           uint64
+	Partition          uint64
+	NewSealedSectorCID cid.Cid
+	Deals              []abi.DealID
+	// TODO: actual type for this
+	Proof []byte
+}
+
+type ProveReplicaUpdatesParams struct {
+	Updates []ReplicaUpdate
+}
+
+// TODO: Should return type be aliased
+func (a Actor) ProveReplicaUpdates(rt Runtime, params *ProveReplicaUpdatesParams) bitfield.BitField {
+	// Validate inputs
+
+	// TODO: check how this gets rendered and fix as needed
+	builtin.RequireParam(rt, len(params.Updates) <= miner3.DeclarationsMax, "too many updates", len(params.Updates), DeclarationsMax)
+
+	// TODO: Any more quick checks before we jump into it?
+	// see extendsectorexpirations for more checks to plagiarize
+
+	// TODO: This is the first time a miner is calling the reward actor directly, but i assume that's fine?
+	// Kinda sad to be doing (and pay this before a bunch of checks, but oh well?
+	var rewRet reward.ThisEpochRewardReturn
+	rewRetCode := rt.Send(builtin.RewardActorAddr, builtin.MethodsReward.ThisEpochReward, nil, big.Zero(), &rewRet)
+	builtin.RequireSuccess(rt, rewRetCode, "failed to check epoch baseline power")
+
+	var powRet power.CurrentTotalPowerReturn
+	powRetCode := rt.Send(builtin.StoragePowerActorAddr, builtin.MethodsPower.CurrentTotalPower, nil, big.Zero(), &rewRet)
+	builtin.RequireSuccess(rt, powRetCode, "failed to check epoch baseline power")
+
+	store := adt.AsStore(rt)
+	var stReadOnly State
+	rt.StateReadonly(&stReadOnly)
+	info := getMinerInfo(rt, &stReadOnly)
+
+	rt.ValidateImmediateCallerIs(append(info.ControlAddresses, info.Owner, info.Worker)...)
+
+	sectors, err := LoadSectors(store, stReadOnly.Sectors)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors array")
+
+	powerDelta := NewPowerPairZero()
+	pledgeDelta := big.Zero()
+
+	// TODO: Pointers? Not pointers?
+	type updateAndSectorInfo struct {
+		update     *ReplicaUpdate
+		sectorInfo *SectorOnChainInfo
+	}
+
+	// Filters out the first wave of invalid updates with quick checks about sector status
+	var prevalidatedUpdates []*updateAndSectorInfo
+
+	var sectorsDeals []market.SectorDeals
+	var sectorsDataSpec []*market.SectorDataSpec
+	for _, update := range params.Updates {
+		// TODO: do we need an assertion that the same sector isn't upgraded twice?
+		// TODO: actual length
+		if len(update.Proof) <= 10 {
+			rt.Log(rtt.INFO, "update proof is wrong length (%d), skipping sector %d", len(update.Proof), update.SectorID)
+			continue
+		}
+
+		if len(update.Deals) <= 0 {
+			rt.Log(rtt.INFO, "must have deals to update, skipping sector %d", update.SectorID)
+			continue
+		}
+
+		// TODO: is there an upper limit on number of deals in a sector (beside the practical limit imposed by sector sizes and min deal size)
+		// TODO Confirm the sector is actually in the stated deadline? This happens later when we try to update partition, but maybe do here first?
```

Comment:
I don't really agree, I don't see why this should be an exception to the error handling principle stated [here](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0019.md#batching-and-error-handling). 

But it's a sufficiently odd error (with no real reason to happen) that I don't feel too strongly -- doing nothing is less work for me :P

Reply:
I was not being precise enough in the wording of the FIP.  As much as possible errors that can't be 100% eliminated by operators should be skipped.  Because deadlines and partitions are effectively immutable operators are going to filter these errors at gas estimation.

Reply:
Yup, sgtm.